### ISSUE TITLE
"highqps" not reaching 95% completion threshold on API priority and fairness ... , blocking gce-master-default

### ISSUE THREAD

**Issue by RobertKielty** [kubernetes/kubernetes#96710]
**Which jobs are flaking**:
gce-cos-master-default

**Which test(s) are flaking**:
Kubernetes e2e suite.[sig-api-machinery] API priority and fairness should ensure that requests can't be drowned out (priority)

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default

**Reason for failure**:

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/flowcontrol.go:76
Nov 18 22:04:59.326: client "highqps": got 94.6% completed requests, want at least 95%
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

**Anything else we need to know**:
- go.k8s.io/triage at present only has data up to 11/11/2020
Three failures here 
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1329227465779843072
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1329209095974031360
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1329180404124487680

**Comment by adtac** [comment]
huh, the highqps client is getting fewer requests in. that's unexpected.

@lavalamp @MikeSpreitzer @yue9944882 what do you think about a lower threshold for the highqps client? the test is intended to test that the lowqps client can still get requests in, so I think we can relax the requirements on the highqps client (say 80% since it's operating at a concurrency far higher than its share). the lowqps client's threshold will remain at 95%.

**Comment by adtac** [comment]
interestingly, I just saw a lowqps flake too: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/96722/pull-kubernetes-e2e-gce-ubuntu-containerd/1329697919141416960/

I'll try to reproduce and investigate this, but I heavily suspect the only solution might be relax both thresholds

**Comment by lavalamp** [comment]
Given that the test cuts off right at the end without waiting for the final request(s) I think you should estimate the expected success percentage assuming a bunch of latency. You could also go the other way and estimate how much longer we need to wait to get to 95%.

Leave a big safety margin!

**Comment by thejoycekung** [comment]
another example of `lowqps` also flaking: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1330746134821343232

**Comment by thejoycekung** [comment]
and just a note that I've noticed this test flaking on a few other jobs as well:
- ci-kubernetes-e2e-ubuntu-gce
- ci-kubernetes-e2e-gci-gce
- ci-kubernetes-e2e-ubuntu-gce-containerd

and triage link: https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=API%20priority#a9c1d8320cab7ba69110

**Comment by marseel** [comment]
Also it's flaking on these tests:
Testgrid: https://k8s-testgrid.appspot.com/sig-release-master-informing#gce-master-scale-correctness

**Comment by RobertKielty** [comment]
Still flaking on https://testgrid.k8s.io/sig-release-master-blocking#gce-ubuntu-master-containerd&exclude-non-failed-tests=&width=20

Here are three failures on commit e3de62298 

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-ubuntu-gce-containerd/1333147517365260288
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-ubuntu-gce-containerd/1332995263807098880
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-ubuntu-gce-containerd/1332873461676118016

**Comment by mborsz** [comment]
ci-kubernetes-e2e-gce-scale-correctness is still flaky with similar error. Most recent failures:
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/1332655775578853376
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/1331568698460213248
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/1331568698460213248

**Comment by adtac** [comment]
We believe https://github.com/kubernetes/kubernetes/pull/96984 might address the true root cause

**Comment by adtac** [comment]
@mborsz @RobertKielty also, AFAICT all of those failures are from before the mitigation (https://github.com/kubernetes/kubernetes/pull/96798) was merged. Testgrid has been green for the last 3 days. Not sure how many passing tests are needed to consider a flaky test to no longer be flaky -- given that it was failing once in every 3-4 tests previously, if we get 8 consecutive passes, can we close this issue?

**Comment by RobertKielty** [comment]
@adtac @mborsz 

This test is still flaking at a 75% threshold on this job 

https://testgrid.k8s.io/sig-release-1.20-blocking#gce-cos-k8sbeta-default&exclude-non-failed-tests=&width=90

Example failure to achieve the 75% threshold.

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-cos-k8sbeta-default/1334787370620293120

Can the test be changed to report on the reason why the thresholds were not met? Is it a timeout issue, a failure to provision a resource? etc. 

We need more detailed signal here in order to know if this is release blocking. Are we failing to meet the required threshold owing to a problem with the code or a problem with the infrastructure?

Also could we have this issue tagged with the PR that were merged to change the threshold if that was done. 

Many thanks for all help on this.

**Comment by adtac** [comment]
>Can the test be changed to report on the reason why the thresholds were not met? Is it a timeout issue, a failure to provision a resource? etc.

We believe the failure is due to extremely high request latency, which could be due to any number of factors (high load on the e2e API server, for example). These are harder to debug because of unreproducability in dev machines. We're disabling the two tests temporarily.

>We need more detailed signal here in order to know if this is release blocking. Are we failing to meet the required threshold owing to a problem with the code or a problem with the infrastructure?

The problem is with the test environment / test parameters, not with the API server itself. The failing test was added as a nice-to-have thing that tests code paths that are already tested elsewhere (in integration tests, unit tests, etc.).

**Comment by pohly** [comment]
Why was this closed? The tests still exist and imply that further work is needed:

https://github.com/kubernetes/kubernetes/blob/2cdcccbc99e7fa62e5930758ff53f1e6577c9461/test/e2e/apimachinery/flowcontrol.go#L102-L109

https://github.com/kubernetes/kubernetes/blob/2cdcccbc99e7fa62e5930758ff53f1e6577c9461/test/e2e/apimachinery/flowcontrol.go#L187-L194

I see several paths forward:
- Remove the tests, with no replacement. Not good for coverage of the feature?
- Rewrite as integration tests, possibly using the new `synctest` Go package for simulating time reliably. Probably the right solution. I tried `synctest` with Go 1.25.1 and a known issue made it unusable, but the current 1.25.2 might work.

In the meantime, replacing `ginkgo.Skip("skipping test until flakiness is resolved")` with the `framework.WithFlaky()` label would be more in line with current best practices - not crucial, but that's how I noticed the tests.

/reopen
